File: internal/relay/info.go
package relay

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/yacchi/backlog-cli/internal/domain"
	jwkutil "github.com/yacchi/backlog-cli/internal/jwk"
)

const relayInfoVersion = 1
const relayInfoDefaultTTL = 600

type relayInfoPayload struct {
	Version       int    `json:"version"`
	RelayURL      string `json:"relay_url"`
	AllowedDomain string `json:"allowed_domain"`
	Space         string `json:"space"`
	Domain        string `json:"domain"`
	IssuedAt      string `json:"issued_at"`
	ExpiresAt     string `json:"expires_at"`
	UpdateBefore  string `json:"update_before,omitempty"`
}

type relayInfoSignature struct {
	Protected string `json:"protected"`
	Signature string `json:"signature"`
}

type relayInfoResponse struct {
	Payload        string               `json:"payload"`
	Signatures     []relayInfoSignature `json:"signatures"`
	PayloadDecoded relayInfoPayload     `json:"payload_decoded"`
}

type relayJWKS struct {
	Keys []relayJWK `json:"keys"`
}

type relayJWK struct {
	Kty string `json:"kty"`
	Crv string `json:"crv"`
	Kid string `json:"kid"`
	X   string `json:"x"`
	D   string `json:"d"`
}

type relayInfoProtected struct {
	Alg string `json:"alg"`
	Kid string `json:"kid"`
}

func (s *Server) handleRelayInfo(w http.ResponseWriter, r *http.Request) {
	allowedDomain := strings.TrimSpace(r.PathValue("domain"))
	if allowedDomain == "" {
		http.Error(w, "domain is required", http.StatusBadRequest)
		return
	}

	tenant, ok := findTenantByAllowedDomain(s.cfg.Server().Tenants, allowedDomain)
	if !ok {
		http.Error(w, "tenant not found", http.StatusNotFound)
		return
	}

	issuedAt := time.Now().UTC()
	ttl := tenant.InfoTTL
	if ttl <= 0 {
		ttl = relayInfoDefaultTTL
	}
	expiresAt := issuedAt.Add(time.Duration(ttl) * time.Second)

	space, backlogDomain := domain.SplitDomain(tenant.AllowedDomain)
	payload := relayInfoPayload{
		Version:       relayInfoVersion,
		RelayURL:      s.buildRelayURL(r),
		AllowedDomain: tenant.AllowedDomain,
		Space:         space,
		Domain:        backlogDomain,
		IssuedAt:      issuedAt.Format(time.RFC3339),
		ExpiresAt:     expiresAt.Format(time.RFC3339),
	}

	payloadJSON, err := json.Marshal(payload)
	if err != nil {
		http.Error(w, fmt.Sprintf("failed to marshal payload: %v", err), http.StatusInternalServerError)
		return
	}
	payloadB64 := base64.RawURLEncoding.EncodeToString(payloadJSON)

	signatures, err := buildRelayInfoSignatures([]byte(payloadB64), tenant.JWKS, tenant.ActiveKeys)
	if err != nil {
		http.Error(w, fmt.Sprintf("failed to sign payload: %v", err), http.StatusInternalServerError)
		return
	}

	resp := relayInfoResponse{
		Payload:        payloadB64,
		Signatures:     signatures,
		PayloadDecoded: payload,
	}

	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(resp)
}

func buildRelayInfoSignatures(payloadB64 []byte, jwksJSON string, activeKeys string) ([]relayInfoSignature, error) {
	activeKeyIDs := splitKeyList(activeKeys)
	if len(activeKeyIDs) == 0 {
		return nil, fmt.Errorf("active_keys is empty")
	}

	var jwks relayJWKS
	if err := json.Unmarshal([]byte(jwksJSON), &jwks); err != nil {
		return nil, fmt.Errorf("invalid jwks: %w", err)
	}

	jwkByKid := make(map[string]relayJWK)
	for _, key := range jwks.Keys {
		if key.Kid != "" {
			jwkByKid[key.Kid] = key
		}
	}

	signatures := make([]relayInfoSignature, 0, len(activeKeyIDs))
	for _, kid := range activeKeyIDs {
		jwk, ok := jwkByKid[kid]
		if !ok {
			return nil, fmt.Errorf("jwks missing key: %s", kid)
		}
		privKey, err := jwkutil.Ed25519PrivateKeyFromJWK(jwk.Kty, jwk.Crv, jwk.Kid, jwk.D)
		if err != nil {
			return nil, err
		}

		protected := relayInfoProtected{
			Alg: "EdDSA",
			Kid: kid,
		}
		protectedJSON, err := json.Marshal(protected)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal protected header: %w", err)
		}

		protectedB64 := base64.RawURLEncoding.EncodeToString(protectedJSON)
		signingInput := protectedB64 + "." + string(payloadB64)
		sig := ed25519.Sign(privKey, []byte(signingInput))

		signatures = append(signatures, relayInfoSignature{
			Protected: protectedB64,
			Signature: base64.RawURLEncoding.EncodeToString(sig),
		})
	}

	return signatures, nil
}

func splitKeyList(value string) []string {
	items := strings.Split(value, ",")
	out := make([]string, 0, len(items))
	for _, item := range items {
		item = strings.TrimSpace(item)
		if item != "" {
			out = append(out, item)
		}
	}
	return out
}

File: packages/relay-core/src/handlers/info.ts
/**
 * Relay info handlers for signed relay information.
 */

import { Hono } from "hono";
import type { RelayConfig, TenantConfig } from "../config/types.js";
import { extractRequestContext } from "../utils/request.js";
import {
  type JWKS,
  base64UrlEncode,
  base64UrlDecode,
  signEd25519,
} from "../utils/crypto.js";

const RELAY_INFO_VERSION = 1;
const RELAY_INFO_DEFAULT_TTL = 600;

/**
 * Relay info payload.
 */
interface RelayInfoPayload {
  version: number;
  name: string;
  relay_url: string;
  issued_at: string;
  expires_at: string;
  update_before?: string;
}

/**
 * Relay info signature.
 */
interface RelayInfoSignature {
  protected: string;
  signature: string;
}

/**
 * Relay info response.
 */
interface RelayInfoResponse {
  payload: string;
  signatures: RelayInfoSignature[];
  payload_decoded: RelayInfoPayload;
}

/**
 * Protected header.
 */
interface ProtectedHeader {
  alg: string;
  kid: string;
}

/**
 * Extended tenant config with JWK fields.
 * Note: TenantConfig already includes these fields, but we keep this for clarity.
 */
type ExtendedTenantConfig = TenantConfig;


/**
 * Compute ETag from JWKS and activeKeys.
 */
async function computeInfoETag(
  jwks: string,
  activeKeys: string
): Promise<string> {
  const data = new TextEncoder().encode(jwks + "|" + activeKeys);
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return (
    '"' +
    hashArray
      .slice(0, 16)
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("") +
    '"'
  );
}



/**
 * Find tenant by name.
 */
function findTenant(
  tenants: TenantConfig[] | undefined,
  name: string
): ExtendedTenantConfig | undefined {
  return tenants?.find(
    (t) => t.name.toLowerCase() === name.toLowerCase()
  );
}

/**
 * Build relay URL from config and request context.
 */
function buildRelayUrl(
  baseUrl: string | undefined,
  reqBaseUrl: string
): string {
  return baseUrl || reqBaseUrl;
}


/**
 * Build signatures for relay info.
 */
async function buildRelayInfoSignatures(
  payloadB64: string,
  jwksJson: string,
): Promise<RelayInfoSignature[]> {
  const jwks: JWKS = JSON.parse(jwksJson);
  if (!jwks.keys || jwks.keys.length === 0) {
    throw new Error("jwks has no keys");
  }

  const signatures: RelayInfoSignature[] = [];

  // Sign with keys[0] (signing key)
  for (const jwk of jwks.keys.slice(0, 1)) {
    const kid = jwk.kid;
    if (!kid) {
    }
    if (!jwk.d) {
      throw new Error(`jwk ${kid} missing private key`);
    }

    const protectedHeader: ProtectedHeader = {
      alg: "EdDSA",
      kid,
    };
    const protectedB64 = base64UrlEncode(JSON.stringify(protectedHeader));
    const signingInput = protectedB64 + "." + payloadB64;

    // Get private key seed
    const seed = base64UrlDecode(jwk.d);

    try {
      const sigBytes = await signEd25519(
        seed,
        new TextEncoder().encode(signingInput)
      );
      signatures.push({
        protected: protectedB64,
        signature: base64UrlEncode(sigBytes),
      });
    } catch {
      // Ed25519 signing not supported in this environment
      // Return empty signatures for now
      throw new Error(
        "Ed25519 signing not supported. Relay info endpoint requires Ed25519 support."
      );
    }
  }

  return signatures;
}

/**
 * Create relay info handlers.
 */
export function createInfoHandlers(config: RelayConfig): Hono {
  const app = new Hono();

  /**
   * GET /v1/relay/tenants/:domain/info - Get signed relay info.
   */
  app.get("/v1/relay/tenants/:name/info", async (c) => {
    const name = c.req.param("name")?.trim();
    if (!name) {
      return c.text("name is required", 400);
    }

    const tenant = findTenant(config.tenants, name);
    if (!tenant) {
      return c.text("tenant not found", 404);
    }

    const jwks = config.jwks;
    if (!jwks) {
      return c.text("server jwks not configured", 500);
    }

    // Compute ETag
    const etag = await computeInfoETag(jwks, "");

    // Check If-None-Match
    const ifNoneMatch = c.req.header("If-None-Match");
    if (ifNoneMatch === etag) {
      return new Response(null, { status: 304 });
    }

    const reqCtx = extractRequestContext(c);
    const issuedAt = new Date();
    const ttl = tenant.info_ttl || RELAY_INFO_DEFAULT_TTL;
    const expiresAt = new Date(issuedAt.getTime() + ttl * 1000);

    const payload: RelayInfoPayload = {
      version: RELAY_INFO_VERSION,
      name: tenant.name,
      relay_url: buildRelayUrl(config.server.base_url, reqCtx.baseUrl),
      issued_at: issuedAt.toISOString(),
      expires_at: expiresAt.toISOString(),
    };

    const payloadJson = JSON.stringify(payload);
    const payloadB64 = base64UrlEncode(payloadJson);

    try {
      const signatures = await buildRelayInfoSignatures(
        payloadB64,
        jwks,
      );

      const response: RelayInfoResponse = {
        payload: payloadB64,
        signatures,
        payload_decoded: payload,
      };

      return c.json(response, 200, {
        ETag: etag,
        "Cache-Control": `public, max-age=${Math.floor(ttl / 2)}`,
      });
    } catch (err) {
      return c.text(`failed to sign payload: ${(err as Error).message}`, 500);
    }
  });

  return app;
}
